README.md
#Hovercrafts

|   |  Max speed | Health  |  Fuel |  Weapon |  Equipment grid |
|---|---|---|---|---|---|
|  Hovercraft |  187 |  500 |  Fossil |  ❌ |  optional |
|  Missilecraft |  115 |  1200 |  Fossil |  Missiles |  optional |
|  Electric Hovercraft |  216 |  250 |  Electric |  ❌ |  8x8 |
|  Laser Hovercraft |  151 |  800 |  Electric |  Laser |  10x10 |

To access Electric and Laser hovercraft, you will need [Laser Tanks Updated](https://mods.factorio.com/mod/laser_tanks_updated)

#Recommended
 - [AAI Programmable Vehicles](https://mods.factorio.com/mod/aai-programmable-vehicles) | Standard hovercraft can be used as a hauler. (Thanks, [Earendel](https://mods.factorio.com/user/Earendel))
 - [Schallfalke's Transport Group](https://mods.factorio.com/mod/SchallTransportGroup/downloads) | Sorts vehicles and vehicle equipment into their own tab.
 - [Portable power equipment](https://mods.factorio.com/mod/Portable_power) | Provides power equipment that consumes fuel while providing a large amount of power.

#Credits
 - [MostlyNumbers](https://mods.factorio.com/user/MostlyNumbers) | Original mod creator
 - [Arcani_Victus](https://mods.factorio.com/user/Arcani_Victus) | Mod continuation
 - [OwnlyMe](https://mods.factorio.com/user/OwnlyMe) | Big collaborator of this mod
 - [snouz](https://mods.factorio.com/user/snouz) | New graphics, new code, bugfixes
 - [daniels1989](https://mods.factorio.com/user/daniels1989) | Compatibility patch


changelog.txt
---------------------------------------------------------------------------------------------------
Version: 1.2.2
Date: 2022-07-30
  Info:
    - Update by Xorimuth
  Bugfixes:
    - Fixed ripples and splashes not being created
    - Fixed hovercraft-collision colliding with water, which caused drifting over water to not work. This was particularly apparent when drifting over a land/water border
    - Fixed trains colliding with walls (https://mods.factorio.com/mod/Hovercrafts/discussion/6181b19499129db2c2577a89)
    - Fixed missing missile turret locale
  Compatibility:
    - Fixed not being able to walk through SE pipes (https://mods.factorio.com/mod/Hovercrafts/discussion/61b19c0e7a19f2772f5262e8)
    - Fixed SE core miners not being able to be placed on core seams
    - Stopped using a hardcoded collision mask (layer-14) which could cause compatibility issues with mods that use the same mask for something else
  Optimization:
    - Improved performance by caching mod settings and installed mod list (https://mods.factorio.com/mod/Hovercrafts/discussion/61f5ca1f1f7f35bd94096d0b)
---------------------------------------------------------------------------------------------------
Version: 1.2.1
Date: 2021-12-07
  Compatibility:
    - Fixed compatibility with IR2 + Krastorio2.
    - Fixed startup crash when Armor Plating was on.
---------------------------------------------------------------------------------------------------
Version: 1.2.0
Date: 2021-12-07
  Info:
    - Update by snouz
  Visuals:
    - Redone stabilized hovercraft visuals (from scratch). Un-mirrored.
    - Added color mask to hovercrafts.
    - Turret improved and size adjusted to match HD/LD resolutions.
    - New HD shadow, including turret shadow.
    - Improved water reflection.
    - Adjusted light position.
    - New technology icons.
    - New icons.
    - New mod thumbnail.
  Bugfixes:
    - Fixed dependency to Laser Tanks. Removed dependency to deprecated Electric Vehicles.
    - Fixed hovercrafts getting stuck in cliffs and rocks.
  Compatibility:
    - Added Industrial Revolution fuel compatibility (thanks to daniels1989).
    - Added Larger Lamps compatibility (thanks to daniels1989).
  Optimization:
    - Images optimized.
    - Files sorted.
    - Code factorised.
---------------------------------------------------------------------------------------------------
Version: 1.1.1
Date: 2021-01-09
  Changes:
    - Updated missile turret with highres version (Massive thanks to Kamsta99!)
---------------------------------------------------------------------------------------------------
Version: 1.1.0
Date: 2020-11-29
  Changes:
    - Updated for Factorio version 1.1.0
  Bugfixes:
    - Resolved an issue involving the laser hovercraft weapon charger not working as intended
---------------------------------------------------------------------------------------------------
Version: 1.0.3
Date: 2020-09-14
  Changes:
    - Resolved hovercraft not colliding with other hovercraft
    - Resolved hovercraft colliding with landmines which also resolved a conflict with the mod Teleporters
    - Resolved a conflict with the mod Train tunnels
---------------------------------------------------------------------------------------------------
Version: 1.0.2
Date: 2020-08-25
  Changes:
    - Actually included the changes intended for 1.0.1(My bad)
    - Resolved some minor issues caused in mod version 0.18.19
    - Added mod check for "justgo" support to resolve an issue with the mod "Quick Item Search"
---------------------------------------------------------------------------------------------------
Version: 1.0.1
Date: 2020-08-25
  Changes:
    - Added the lamp to the list of entities the hovercrafts shouldn't collide with
---------------------------------------------------------------------------------------------------
Version: 1.0.0
Date: 2020-08-14
  Changes:
    - Updated for Factorio version 1.0.0
---------------------------------------------------------------------------------------------------
Version: 0.18.25
Date: 2020-07-22
  Bugfixes:
    - Resolved an issue involving Schallfalke's Transport group mod
  Changes:
    - Missilecraft: Increased rocket launcher turret range (30 to 36)
    - Missilecraft: Reduced rocket launcher turret cooldown (300 to 120)
---------------------------------------------------------------------------------------------------
Version: 0.18.24
Date: 2020-07-18
  Bugfixes:
    - Resolved an issue involving the recently added icon mipmaps
  Locale:
    - Updated Russian translation (Thanks Va7ya)
---------------------------------------------------------------------------------------------------
Version: 0.18.23
Date: 2020-07-06
  Bugfixes:
    - Resolved an issue with logistic robots crashing into trains
---------------------------------------------------------------------------------------------------
Version: 0.18.22
Date: 2020-07-05
  Bugfixes:
    - Resolved an issue regarding mod check for the lcraft weapon charger
---------------------------------------------------------------------------------------------------
Version: 0.18.21
Date: 2020-07-05
  Bugfixes:
    - Reworked drifting so you should no longer get stuck in cliffs
    - Fix conflict with lovely_santa's Train constuction site mod
  Changes:
    - General
    - Added compatibility for Bob's vehicle equipment mod
    - Added support for lovely_santa's Train construction site mod's transport-logistics tab
    - Lasercraft: Added unique ammo charger for laser hovercraft's laser gun
---------------------------------------------------------------------------------------------------
Version: 0.18.20
Date: 2020-02-03
  Bugfixes:
    - Added support for "Just go!"
    - The collision box of the hovercraft won't kill you anymore when the hovercraft crashes

prototypes/entity.lua
local hcraft_remnants = table.deepcopy(data.raw.corpse["car-remnants"])
hcraft_remnants.name = "hovercraft-remnants"
hcraft_remnants.animation.layers[1].filename = HCGRAPHICS .. "entity/hovercraft/remnants/hovercraft-remnants.png"
hcraft_remnants.animation.layers[1].hr_version.filename = HCGRAPHICS .. "entity/hovercraft/remnants/hr-hovercraft-remnants.png"
data:extend({hcraft_remnants})

-- collision box
local collision = table.deepcopy(data.raw.car.car)
  collision.name = "hovercraft-collision"
  collision.order = "hovercraft-collision"
  collision.collision_box = {{-1.5, -1.5}, {1.5, 1.5}}
  collision.collision_mask = { "player-layer" }  -- Will be replaced by custom collision layer in data-final-fixes
  collision.animation = {filename = "__core__/graphics/empty.png", size = 1, direction_count = 1}
  collision.turret_animation = nil
  collision.light_animation = nil
  collision.light = nil
  collision.water_reflection = nil
data:extend({collision})

local shadow_distance = 11
--------------------------------------------------------------------------------------------------------------------
local hcraft_entity = table.deepcopy(data.raw.car.car)
hcraft_entity.name = "hcraft-entity"
hcraft_entity.icon = HCGRAPHICS .. "icons/hovercraft_icon.png"
hcraft_entity.icon_size = 64
hcraft_entity.corpse = "hovercraft-remnants"
hcraft_entity.braking_power = "1200kW"
hcraft_entity.consumption = "250kW"
hcraft_entity.selection_box = {{-1, -1.2}, {1, 1.2}}
hcraft_entity.collision_box = {{-0.7, -0.9}, {0.7, 0.9}}
hcraft_entity.effectivity = 1.3
hcraft_entity.max_health = 500
hcraft_entity.guns = {}
hcraft_entity.terrain_friction_modifier = 0
hcraft_entity.rotation_speed = 0.0060
hcraft_entity.tank_driving = true
hcraft_entity.weight = 2500
hcraft_entity.minable = {mining_time = 0.5, result = "hcraft-entity"}
hcraft_entity.has_belt_immunity = true
hcraft_entity.collision_mask = { "player-layer" }  -- Will be replaced by custom collision layer in data-final-fixes
hcraft_entity.resistances = {
  { type = "fire",      decrease = 7.5, percent = 30 },
  { type = "physical",  decrease = 7.5, percent = 30 },
  { type = "impact",    decrease = 40,  percent = 65 },
  { type = "explosion", decrease = 7.5, percent = 35 },
  { type = "acid",      decrease = 0,   percent = 35 }
}
hcraft_entity.stop_trigger = {
  { type = "play-sound", sound = { { filename = "__base__/sound/car-breaks.ogg", volume = 0.0 } } }
}
hcraft_entity.animation = {
  layers = {
    {
      filename = HCGRAPHICS .. "entity/hovercraft/hovercraft-base.png",
      priority = "high",
      line_length = 8,
      size = 128,
      max_advance = 0.2,
      direction_count = 64,
      scale = 1,
      hr_version = {
        filename = HCGRAPHICS .. "entity/hovercraft/hr-hovercraft-base.png",
        priority = "high",
        line_length = 8,
        size = 256,
        max_advance = 0.2,
        direction_count = 64,
        scale = 0.5,
      }
    },
    {
      filename = HCGRAPHICS .. "entity/hovercraft/hovercraft-mask.png",
      priority = "low",
      line_length = 8,
      size = 128,
      max_advance = 0.2,
      direction_count = 64,
      scale = 1,
      apply_runtime_tint = true,
      hr_version = {
        filename = HCGRAPHICS .. "entity/hovercraft/hr-hovercraft-mask.png",
        priority = "low",
        line_length = 8,
        size = 256,
        max_advance = 0.2,
        direction_count = 64,
        scale = 0.5,
        apply_runtime_tint = true,
      }
    },
    {
      filename = HCGRAPHICS .. "entity/hovercraft/hovercraft-shadow.png",
      line_length = 8,
      size = 128,
      max_advance = 0.2,
      direction_count = 64,
      shift = util.by_pixel(shadow_distance, shadow_distance),
      draw_as_shadow = true,
      hr_version = {
        filename = HCGRAPHICS .. "entity/hovercraft/hr-hovercraft-shadow.png",
        line_length = 8,
        size = 256,
        max_advance = 0.2,
        direction_count = 64,
        scale = 0.5,
        shift = util.by_pixel(shadow_distance, shadow_distance),
        draw_as_shadow = true,
      }
    },
  }
}
hcraft_entity.light_animation = {
  filename = HCGRAPHICS .. "entity/hovercraft/hovercraft-light.png",
  priority = "low",
  blend_mode = "additive",
  draw_as_glow = true,
  size = 128,
  line_length = 8,
  direction_count = 64,
  hr_version =
  {
    filename = HCGRAPHICS .. "entity/hovercraft/hr-hovercraft-light.png",
    priority = "low",
    blend_mode = "additive",
    draw_as_glow = true,
    size = 256,
    line_length = 8,
    direction_count = 64,
    scale = 0.5,
  }
}
hcraft_entity.water_reflection = {
  pictures = {
    filename = HCGRAPHICS .. "entity/hovercraft/hovercraft-water-reflection.png",
    width = 26,
    height = 26,
    shift = util.by_pixel(0, 15),
    variation_count = 1,
    scale = 5
  },
  rotate = true,
  orientation_to_variation = false
}
hcraft_entity.turret_animation = nil
data:extend({hcraft_entity})

--------------------------------------------------------------------------------------------------------------------
if mcraft_activated then
  -- Mcraft entity
  local mcraft_entity = table.deepcopy(data.raw.car["hcraft-entity"])
  mcraft_entity.name = "mcraft-entity"
  mcraft_entity.braking_power = "1500kW"
  mcraft_entity.consumption = "450kW"
  mcraft_entity.effectivity = 1.1
  mcraft_entity.max_health = 1200
  mcraft_entity.rotation_speed = 0.0045
  mcraft_entity.tank_driving = true
  mcraft_entity.immune_to_tree_impacts = true
  mcraft_entity.immune_to_rock_impacts = true
  mcraft_entity.weight = 10000
  mcraft_entity.minable = {mining_time = 0.5, result = "mcraft-entity"}
  mcraft_entity.resistances = {
    { type = "fire",      decrease = 10, percent = 55 },
    { type = "physical",  decrease = 10, percent = 55 },
    { type = "impact",    decrease = 30, percent = 65 },
    { type = "explosion", decrease = 10, percent = 65 },
    { type = "acid",      decrease = 0,  percent = 55 }
  }
  mcraft_entity.burner = {
    fuel_category = "chemical",
    effectivity = 1,
    fuel_inventory_size = 2,
  }
  mcraft_entity.guns = {"hovercraft-missile-turret"}
  mcraft_entity.turret_animation = {
    layers = {
      {
        filename = HCGRAPHICS .. "entity/turret/hovercraft-missile-turret.png",
        line_length = 8,
        size = 64,
        direction_count = 64,
        shift = util.by_pixel(0, -27),
        hr_version = {
          filename = HCGRAPHICS .. "entity/turret/hr-hovercraft-missile-turret.png",
          line_length = 8,
          size = 128,
          direction_count = 64,
          shift = util.by_pixel(0, -27),
          scale = 0.5,
        }
      },
      {
        filename = HCGRAPHICS .. "entity/turret/hovercraft-missile-turret-shadow.png",
        line_length = 8,
        size = 64,
        direction_count = 64,
        shift = util.by_pixel(30+shadow_distance, 8+shadow_distance),
        draw_as_shadow = true,
        hr_version = {
          filename = HCGRAPHICS .. "entity/turret/hr-hovercraft-missile-turret-shadow.png",
          line_length = 8,
          size = 128,
          direction_count = 64,
          shift = util.by_pixel(30+shadow_distance, 8+shadow_distance),
          draw_as_shadow = true,
          scale = 0.5,
        }
      },
    }
  }
  mcraft_entity.turret_rotation_speed = 0.40 / 60
  mcraft_entity.turret_return_timeout = 300

  local mcraft_gun = table.deepcopy(data.raw.gun["vehicle-machine-gun"])
  mcraft_gun.name = "hovercraft-missile-turret"
  mcraft_gun.icon = HCGRAPHICS .. "icons/hovercraft-missile-turret-icon.png"
  mcraft_gun.icon_size = 64
  mcraft_gun.icon_mipmaps = 0
  mcraft_gun.order = "d[rocket-launcher]"
  mcraft_gun.attack_parameters = {
    type = "projectile",
    ammo_category = "rocket",
    cooldown = 120, --60, --300,
    movement_slow_down_factor = 0.9,
    projectile_center = {-0.17, 0},
    projectile_creation_distance = 0.6,
    range = 36,
    sound = {
      {
        filename = "__base__/sound/fight/rocket-launcher.ogg",
        volume = 0.7
      }
    }
  }
  data:extend({mcraft_entity, mcraft_gun})
end

--------------------------------------------------------------------------------------------------------------------
if ecraft_activated then
  local ecraft_entity = table.deepcopy(data.raw.car["hcraft-entity"])
  ecraft_entity.name = "ecraft-entity"
  ecraft_entity.icon = HCGRAPHICS .. "icons/hovercraft_ecraft_icon.png"
  ecraft_entity.icon_size = 64
  ecraft_entity.braking_power = "1000kW"
  ecraft_entity.consumption = "6MW"
  ecraft_entity.effectivity = 0.11
  ecraft_entity.max_health = 250
  ecraft_entity.rotation_speed = 0.0075
  ecraft_entity.weight = 1500
  ecraft_entity.minable = {mining_time = 0.5, result = "ecraft-entity"}
  ecraft_entity.equipment_grid = "ecraft-equipment"
  ecraft_entity.sound_no_fuel = {
    {
      filename = "__Hovercrafts__/audio/no-energy.ogg",
      volume = 0.4
    }
  }
  ecraft_entity.working_sound = {
    sound =
    {
      filename = "__Hovercrafts__/audio/vehicle-motor.ogg",
      volume = 0.5
    },
    match_speed_to_activity = false
  }
  ecraft_entity.burner =
  {
    effectivity = nil,
    fuel_inventory_size = 0,
  }
  data:extend({ecraft_entity})
end

--------------------------------------------------------------------------------------------------------------------
if lcraft_activated then
  local lcraft_entity = table.deepcopy(data.raw.car["hcraft-entity"])
  lcraft_entity.name = "lcraft-entity"
  lcraft_entity.icon = HCGRAPHICS .. "icons/hovercraft_lcraft_icon.png"
  lcraft_entity.icon_size = 64
  lcraft_entity.effectivity = 0.20
  lcraft_entity.max_health = 800
  lcraft_entity.rotation_speed = 0.0050
  lcraft_entity.weight = 7500
  lcraft_entity.minable = {mining_time = 0.5, result = "lcraft-entity"}
  lcraft_entity.equipment_grid = "lcraft-equipment"
  lcraft_entity.immune_to_tree_impacts = true
  --lcraft_entity.immune_to_rock_impacts = true
  lcraft_entity.burner =
  {
    effectivity = nil,
    fuel_inventory_size = 0,
  }
  lcraft_entity.braking_power = "1250kW"
  lcraft_entity.consumption = "8MW"
  lcraft_entity.sound_no_fuel = {
    {
      filename = "__Hovercrafts__/audio/no-energy.ogg",
      volume = 0.4
    }
  }
  lcraft_entity.working_sound = {
    sound = {
      filename = "__Hovercrafts__/audio/vehicle-motor.ogg",
      volume = 0.5
    },
    match_speed_to_activity = false
  }
  lcraft_entity.resistances = {
    { type = "fire", decrease = 7.5, percent = 30 },
    { type = "physical", decrease = 7.5, percent = 30 },
    { type = "impact", decrease = 40, percent = 75 },
    { type = "explosion", decrease = 7.5, percent = 35 },
    { type = "acid", decrease = 0, percent = 35 }
  }
  lcraft_entity.guns = {"vehicle-laser-gun"}
  lcraft_entity.turret_rotation_speed = 0.35 / 60
  lcraft_entity.turret_animation = data.raw.car.car.turret_animation
  data:extend({lcraft_entity})
end